File: VioletMarketplace/src/main/java/com/skilldistillery/marketplace/services/SearchServiceImpl.java
package com.skilldistillery.marketplace.services;

import com.skilldistillery.marketplace.dto.SearchResults;
import com.skilldistillery.marketplace.entities.Token;
import com.skilldistillery.marketplace.entities.User;
import com.skilldistillery.marketplace.enums.AccountStatus;
import com.skilldistillery.marketplace.enums.Status;
import com.skilldistillery.marketplace.repositories.TokenRepository;
import com.skilldistillery.marketplace.repositories.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Set;


@Service
public class SearchServiceImpl implements SearchService {
    public static final String ALL_QUERY = "all";


    @Autowired
    private TokenRepository tokenRepo;

    @Autowired
    private UserRepository userRepo;

    public SearchResults search(String query) {
        SearchResults searchResults = new SearchResults();

        if(query.trim().equalsIgnoreCase(ALL_QUERY)) {
            Set<Token> tokenResults = tokenRepo.findByStatus(Status.AVAILABLE);
            List<User> userResults = userRepo.findByAccountStatus(AccountStatus.ACTIVE);
            searchResults.setUsers(userResults);
            searchResults.setTokens(tokenResults);
            return searchResults;
        }

        Set<Token> tokenResults = tokenRepo.findByNameOrDescriptionIgnoreCase(query);
        List<User> userResults = userRepo.findByUsernameOrDisplayNameIgnoreCase(query);

        searchResults.setTokens(tokenResults);
        searchResults.setUsers(userResults);

        return searchResults;
    }
}

File: VioletMarketplace/src/main/java/com/skilldistillery/marketplace/services/SearchServiceImpl.kt
package com.skilldistillery.marketplace.services

import com.skilldistillery.marketplace.dtos.SearchResults
import com.skilldistillery.marketplace.dtos.toCommonDTO
import com.skilldistillery.marketplace.entities.Token
import com.skilldistillery.marketplace.entities.User
import com.skilldistillery.marketplace.repositories.TokenRepository
import com.skilldistillery.marketplace.repositories.TrieRepository
import com.skilldistillery.marketplace.repositories.UserRepository
import org.springframework.stereotype.Service

@Service
class SearchServiceImpl(private val userRepo: UserRepository, private val tokenRepo: TokenRepository, private val trieRepo: TrieRepository) : SearchService {
    // TODO removed all check from function, need to adjust accordingly
    override fun search(query: String): SearchResults {
        val trieResults = trieRepo.trie.search(query)
        trieResults.forEach { println("trie results: $it") }

        val userResults = trieResults.filter { it.type == "User" }.map { userRepo.findById(it.id).get().toCommonDTO() }.toMutableSet()
        val tokenResults = trieResults.filter { it.type == "Token" }.map { tokenRepo.findById(it.id).get().toCommonDTO() }.toMutableSet()
        userResults.forEach { println("user results: $it") }
        tokenResults.forEach { println("token results: $it") }
        return SearchResults(userResults, tokenResults)
    }
}
